feat(alu): Implement Mul64Chip to prove I32Mul64 operations

perseus34x opened the pull request:
This operation takes two 32-bit signed integers, multiplies them as 64-bit values, and pushes the full 64-bit result back to the stack as two 32-bit words (high and low).
The proof verification is handled within the Mul64Chip by constraining the following logic:

1. Zero Extension: The two 32-bit input operands (b and c) are zero-extended to 64-bit values. 
2. 32x32 Bit Multiplication: To prove the multiplication in a finite field, the operation is broken down using a schoolbook multiplication method:



## PR Checklist

- [X] Added Tests
- [X] Added Documentation
- [X] Breaking changes

nagatoism committed:
feat: i64 memory access

Commit:
support i32Mul64

Commit:
Merge remote-tracking branch 'origin/change-executor-to-rwasm-native' into perseus34x/support-i64mul

# Conflicts:
#	crates/rwasm/executor/src/events/instr.rs

Commit:
support i32Mul64

nagatoism commented on crates/rwasm/machine/src/alu/mul64/mod.rs:
```diff
@@ -0,0 +1,414 @@
+            },
+        ]);
+
+        let mut product = [0u32; LONG_WORD_SIZE];
```
Convert to 64bits do multiplication and convert the result to an array.
loop not needed.

nagatoism commented on crates/rwasm/machine/src/alu/mul64/mod.rs:
```diff
@@ -0,0 +1,414 @@
+        builder.assert_bool(local.c_sign_extend);
+
+        // Sign extend b and c
+        let (b, c) = {
```
the result can be constrained by the Mul chip. Not need to redo the multiplication.

Commit:
refactor i32Mul64 and

Commit:
cargo update for rwasm

Commit:
fix nb_rows

Commit:
rename a to a_lo for I64event and fix send64_instruction

Commit:
add constraints in cpu for res_lo and res_li and sp and next_p

Commit:
remove problematic constraint

Commit:
remove problematic constraint

djadjka commented on crates/rwasm/machine/src/alu/mul64/mod.rs:
```diff
@@ -0,0 +1,435 @@
+    pub b_sign_extend: T,
+    pub c_sign_extend: T,
+    pub is_mul64: T,
+    pub is_real: T,
```
not sure that you need is_real and is_mul64, pls use only one form it

djadjka commented on crates/rwasm/machine/src/alu/mul64/mod.rs:
```diff
@@ -0,0 +1,435 @@
+        }
+
+        cols.product = product.map(F::from_canonical_u32);
+        cols.a_lo = Word(a_lo_word.map(F::from_canonical_u8));
```
i think you can use event.a_lo.into() here

perseus34x replied:
and for a_hi also

djadjka commented on crates/rwasm/machine/src/alu/mul64/mod.rs:
```diff
@@ -0,0 +1,435 @@
+        );
+
+        // Sign extension calculation
+        builder.assert_eq(local.b_sign_extend, local.b_msb);
```
do we need local._b_sign_extend_ and local._b_msb_ if there are eq ?

Commit:
fix memory

Commit:
fix memory constraints

Commit:
refactor code

Commit:
Merge branch 'change-executor-to-rwasm-native' into perseus34x/support-i64mul

# Conflicts:
#	crates/rwasm/executor/src/air.rs
#	crates/rwasm/executor/src/artifacts/rv32im_costs.json
#	crates/rwasm/machine/src/rwasm/mod.rs
#	crates/rwasm/machine/src/shape/mod.rs
#	crates/rwasm/prover/src/rwasmtest/mod.rs

Commit:
refactor code

Commit:
refactor code

Commit:
refactor code

Commit:
change the multiplication approach as follows:

Equivalent to the expression (int64_t)lhs * (int64_t)rhs in high-level languages where lhs and rhs are u32.

for example
let b = 12345u32;
let c = 54321u32;
let correct_res = (b as i64).wrapping_mul(c as i64);

Commit:
cargo update

Commit:
remove product column and redundant constraints

djadjka commented on crates/rwasm/machine/src/alu/mul64/mod.rs:
```diff
@@ -0,0 +1,374 @@
+        let mut b = b_word.to_vec();
+        let mut c = c_word.to_vec();
+        // For I32Mul64 we now interpret b and c as u32 and zero-extend them to 64 bits.
+        b.resize(LONG_WORD_SIZE, 0);
```
you don't need this resize

djadjka commented on crates/rwasm/machine/src/alu/mul64/mod.rs:
```diff
@@ -0,0 +1,374 @@
+        let b_word = event.b.to_le_bytes();
+        let c_word = event.c.to_le_bytes();
+
+        let mut b = b_word.to_vec();
```
you don't need cast to vec

djadjka commented on crates/rwasm/machine/src/alu/mul64/mod.rs:
```diff
@@ -0,0 +1,374 @@
+
+        cols.a_lo = event.a_lo.into();
+        cols.a_hi = event.a_hi.into();
+        cols.b = Word(b_word.map(F::from_canonical_u8));
```
just use events.b.into()

djadjka commented on crates/rwasm/machine/src/alu/mul64/mod.rs:
```diff
@@ -0,0 +1,374 @@
+        let local = main.row_slice(0);
+        let local: &Mul64Cols<AB::Var> = (*local).borrow();
+        let base = AB::F::from_canonical_u32(1 << 8);
+        let zero: AB::Expr = AB::F::zero().into();
```
AB::Expr::zero()

djadjka commented on crates/rwasm/machine/src/alu/mul64/mod.rs:
```diff
@@ -0,0 +1,374 @@
+
+        // Zero-extend b and c from 32 bits to 64 bits (unsigned semantics).
+        let (b, c) = {
+            let mut b: Vec<AB::Expr> = vec![zero.clone(); LONG_WORD_SIZE];
```
You don't need to provide type here, rust already know that it will be Vec\<AB::Expr\>

djadjka commented on crates/rwasm/machine/src/alu/mul64/mod.rs:
```diff
@@ -0,0 +1,374 @@
+        let (b, c) = {
+            let mut b: Vec<AB::Expr> = vec![zero.clone(); LONG_WORD_SIZE];
+            let mut c: Vec<AB::Expr> = vec![zero.clone(); LONG_WORD_SIZE];
+            for i in 0..LONG_WORD_SIZE {
```
_for i in 0..WORD_SIZE_ and you don't need add if else

Commit:
refactore code

Commit:
Optimize Mul64Chip by restricting the multiplication grid to 4x4 for u32 inputs, eliminating redundant constraints on zero-extended upper bytes.

perseus34x commented:
It is implemented in AddMul64Chip.
